[ENH][BEP028] Specification update for BEP028 BIDS-Prov

bclenet opened the pull request:
# BEP 028: Under review

This BEP is currently under review: https://github.com/bids-standard/bids-specification/discussions/2405

Please post significant discussion to that thread. Minor issues, such as typos or clarifications, may be posted as suggestions to this PR. These suggestions may be accepted or rejected without discussion, so do not make substantial proposals in this way.

# Original post

This is a work in progress PR proposing a specification update for BEP028 BIDS-Prov.

## Todo lists

- https://github.com/bids-standard/BEP028_BIDSprov/issues/167
- https://github.com/bids-standard/BEP028_BIDSprov/issues/168

## Blocking points

- https://github.com/bids-standard/BEP028_BIDSprov/issues/169
- https://github.com/bids-standard/BEP028_BIDSprov/issues/172

## Associated BIDS examples

- [[BEP028] provenance - custom script using Nilearn #524](https://github.com/bids-standard/bids-examples/pull/524)
- [[BEP028] provenance - study dataset + manual annotation #519](https://github.com/bids-standard/bids-examples/pull/519)
- [[BEP028] provenance - fMRIPrep preprocessing #502](https://github.com/bids-standard/bids-examples/pull/502)
- [[BEP028] provenance - SPM preprocessing #497](https://github.com/bids-standard/bids-examples/pull/497)
- [[BEP028] provenance - heudiconv conversion #496](https://github.com/bids-standard/bids-examples/pull/496)
- [[BEP028] provenance - dcm2niix conversion #494](https://github.com/bids-standard/bids-examples/pull/494)

## Shortcuts

- RTD Preview: https://bids-specification--2099.org.readthedocs.build/en/2099/modality-agnostic-files/provenance.html

bclenet commented:
Hi @yarikoptic,

As part of this PR (intagration of BEP028 in the BIDS specification), I'm adding the BIDS-Prov context into the repo. For now it's under : `src/modality-agnostic-files/provenance-context.json` but do you see a better place where to add it ? Thanks !

bclenet committed:
Entity Ids are BIDS URIs

bclenet committed:
Remove GeneratedByProv -> integrated into GeneratedBy for dataset level provenance description

bclenet committed:
Entities allowed in prov-*_ent.json files

bclenet committed:
Remark

bclenet committed:
Entities allowed in prov-*_ent.json files

bclenet committed:
Removing GeneratedByProv

bclenet committed:
GeneratedBy field for Entities allow arrays

bclenet committed:
fMRIPrep example

bclenet committed:
Markdown lint

bclenet committed:
Merge branch 'master' into BEP028_spec

bclenet committed:
BIDS-Prov citation + Provenance example links

bclenet committed:
Correcting typos

bclenet committed:
Links from snippets to bids-examples

bclenet committed:
Links from snippets to bids-examples

bclenet committed:
Provenance graph for definitions

bclenet committed:
Provenance example graph with neuro context

yarikoptic commented on src/schema/rules/files/common/tables.yaml:
```diff
@@ -5,12 +5,14 @@ participants:
   extensions:
     - .tsv
     - .json
+
```
there should ideally be no unrelated changes in the PR

yarikoptic commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+-->
+{{ MACROS___make_subobject_table("metadata.Environments.items") }}
+
+!!! bug
+    TODO: Environment not currently defined in the context
```
what to be done about that?

bclenet replied:
Is there an IRI from PROV-O or from any other ontology we should map with `Enviroment` ?

satra replied:
`Environment` is a `prov:Entity`

cmaumet replied:
Just realized now that actually in BIDS-Prov, it's easy to differentiate as "Environments" and "Entities" arrays as they are defined using different keywords (although in RDF they both map to prov:Entity) so I think we can keep as is. 

Sorry for the noise!

yarikoptic commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+!!! example "`Activity`, `Environment`, `Software` naming examples"
+    - `bids:ds001734:prov#conversion-xfMMbHK1`: a conversion `Activity` described inside the `ds001734` dataset;
+    - `bids::prov#fedora-uldfv058`: a Fedora based `Environment` described inside the current dataset.
+    - `bids:derivatives:prov#fmriprep-r4kzzMt8`: the fMRIPrep `Software` described inside the `derivatives` dataset.
```
I feel we would need some very explicit statement about uniqueness of Ids, and enforce that in validator somehow.

```suggestion
    - `bids:derivatives:prov#fmriprep-r4kzzMt8`: the fMRIPrep `Software` described inside the `derivatives` dataset.

An `Id` must be unique, as to identify an instance and thus REQUIRED to be different across entries and different instances MUST have different `Id`s.
```

yarikoptic commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+
+!!! example
+    The following graph represents examples of links between provenance records. In this example, `Entities` *sub-001_brainmask.nii* and *sub-001_T1w.nii* represent files used by `Activity` *Brain extraction* to generate another file represented by `Entity` *sub-001_T1w_preproc.nii*. `Activities` *Brain extraction* and *Move to MNI* were associated with the `Software` *FSL* and used `Environment` ...
+    ![](../images/provenance_definitions_graph.png)
```
is there a way to ship an original graph or ideally a collection of prov files + script to produce such a graph?

bclenet replied:
For now, the `bids_prov.merge` and `bids_prov.visualize` python modules are available in https://github.com/bids-standard/BEP028_BIDSprov/, to produce such a graph from a collection of prov files.
Would you like to have the prov files for this minimal example inside bids examples ?

yarikoptic commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+
+## Provenance in sidecar JSON files
+
+Provenance metadata CAN be stored inside the sidecar JSON of any BIDS file (or BIDS-Derivatives file) it applies to.
```
```suggestion
Provenance metadata CAN be stored inside the sidecar JSON of any BIDS file it applies to.
```

I do not see a point of "or" here.

yarikoptic commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+Provenance metadata CAN be stored inside the sidecar JSON of any BIDS file (or BIDS-Derivatives file) it applies to.
+This metadata only describes the provenance of the BIDS file.
+
+Any sidecar JSON file CAN include the following keys:
```
```suggestion
Any sidecar JSON file MAY include the following keys:
```

per https://bids-specification--2099.org.readthedocs.build/en/2099/common-principles.html#definitions where no `CAN` is defined . Change here and elsewhere in the text

yarikoptic commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+   }
+) }}
+
+No other field is allowed to describe provenance inside sidecar JSON files.
```
ATM sidecar is "open" for defining any other field and some software might...   hence I would just remove this

```suggestion
```

related
- https://github.com/bids-standard/bids-specification/issues/637
- https://github.com/bids-standard/bids-2-devel/issues/80

yarikoptic commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+No other field is allowed to describe provenance inside sidecar JSON files.
+
+!!! note
+    Not defining the `SidecarGenearatedBy` field means that the sidecar JSON was generated by the `Activity` described in the `GenearatedBy` field.
```
```suggestion
    Not defining the `SidecarGenearatedBy` field means that the sidecar JSON was generated by the same `Activity` as the data file itself and described in the `GenearatedBy` field.
```
to clarify?

yarikoptic commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+
+## Provenance at dataset level
+
+Provenance metadata CAN be stored inside the `dataset_description.json` of any BIDS dataset (or BIDS-Derivatives dataset) it applies to.
```
```suggestion
Provenance metadata MAY be stored inside the `dataset_description.json` of any BIDS dataset (or BIDS-Derivatives dataset) it applies to.
```

yarikoptic commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+   }
+) }}
+
+The `dataset_description.json` file of a BIDS-Derivatives dataset MUST include the following key to describe provenance:
```
this is from the "common-principles", so  better be referred to it?

bclenet replied:
Do you suggest to remove all metadata tables from the _Provenance at dataset level_ section because these metadata are already listed in `src/modality-agnostic-files/dataset-description.json` ?

cmaumet replied:
This is not duplication, the suggestion is to move all the provenance-related info (including how to write our GeneratedBy) from common-principles into Provenance. See https://github.com/bids-standard/bids-specification/pull/2099#discussion_r2352530874

yarikoptic reviewed (changes requested):
some initial notes

satra reviewed (commented):
looks very nice @bclenet - left some comments which should be very lightweight to address.

bclenet committed:
Merge branch 'master' into BEP028_spec

bclenet committed:
Remove unrelated changes

bclenet committed:
Typo and replace CAN by MAY

bclenet committed:
Examples at the begining of subsections

bclenet committed:
Uniqueness of ids

bclenet commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+
+!!! Caution
+    These files MUST not contain `Entity` records describing data files that are available in the dataset. Use sidecar JSON files instead for this purpose (see [Provenance in sidecar JSON files](#provenance-in-sidecar-json-files)).
+    These files MUST not contain `Entity` records describing the current dataset. Use the `GeneratedBy` field of the `dataset_description.json` files instead for this purpose (see [Provenance at dataset level](#provenance-at-dataset-level)).
```
This is odd indeed, I rephrased the sentence.

bclenet commented on src/modality-agnostic-files/provenance.md:
````diff
@@ -0,0 +1,485 @@
+    }
+) }}
+
+The following provenance record is defined in `prov/prov-dcm2niix_soft.json`, so its `Id` SHOULD start with `bids:<dataset>:prov#` or `bids::prov#`.
+
+```JSON
+{
+    "Software": [
+        {
+            "Id": "bids::prov#dcm2niix-70ug8pl5",
+            "Label": "dcm2niix",
+            "Version": "v1.1.3"
+        }
+    ]
+}
+```
````
I'll rephrase this.
We simply give an example illustrating the _Consistency and uniqueness of Ids_ section here. There is indeed no correlation between the file `prov/prov-dcm2niix_soft.json` and the fact that the Id starts with `bids:<dataset>:prov#`

bclenet committed:
Minimal example > consistency of Ids

yarikoptic reviewed (changes requested):
Next batch of comments and recommendations

bclenet commented on src/common-principles.md:
```diff
@@ -312,10 +312,11 @@ Derivatives can be stored/distributed in two ways:
+    the `Name` field of the first `GeneratedBy` object in `dataset_description.json`,
+    optionally followed by a hyphen and a suffix.
+    That is, if a directory `<dataset>/derivatives/<pipeline>[-<variant>]/` is used, the first
+    `GeneratedBy` object should have a `Name` of `<pipeline>`.
```
This seems to be related to a problem while merging latest changes from the master branch.
I'll stick to what is in the master branch.

bclenet committed:
Issue with dataset naming

bclenet committed:
Simplification in definitions

bclenet commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+**Provenance record**: provenance metadata consists in sets of records. These provenance records can be of 4 types:
+
+-   `Activity`: Activities represent the transformations that have been applied to the data.
+-   `Entity`: Entities represent input or output data for Activities.
```
In this definitions section, I did not apply the exact changes you suggested (but it still goes in the same direction). Please let me know if its ok.

bclenet committed:
Deidentification

bclenet committed:
Merge branch 'master' into BEP028_spec

bclenet committed:
Removing prefix all

bclenet committed:
Remove unrelated changes

cmaumet commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+
+## Minimal example
+
+Here is a comprehensive example that considers the following dataset:
```
Hey @satra! Can you give us some more insight on what you mean by "study-level organization". Is this a change in the BIDS spec? Or do you mean a re-organization of the files within a BIDS folder for a given study.

satra replied:
here is the relevant doc: https://bids-specification.readthedocs.io/en/stable/common-principles.html#study-dataset this was merged with the recent release.

bclenet replied:
This PR will allow to describe provenance of any BIDS dataset (may it be in the dataet_description.json, in the prov directory, or in a sidecar). This applies to a study dataset as well as to its nested BIDS datasets. Therefore, @cmaumet and I think there is nothing specific to add about study-level inside the Provenance section of the specification.

bclenet committed:
Sidecargeneratedby precision

satra commented on src/images/provenance_definitions_graph.png:
@bclenet - wondering if these could be created with mermaid markdown? this should not hold things up, but may be easier to embed and render if possible.

satra commented on src/modality-agnostic-files/provenance.md:
````diff
@@ -0,0 +1,487 @@
+        ]
+    }
+    ```
+    This snippet is an extract of the following comprehensive example: [Provenance records for fMRI preprocessing using `fMRIPrep`](https://github.com/bclenet/bids-examples/tree/BEP028_fmriprep/provenance_fmriprep).
````
if we are going to point to a separate repo, we should note that it will have to be maintained after the PR is merged. perhaps something to think about and consider whether these examples are transitioned into bids examples repo and linked here.

satra commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,485 @@
+            {
+                "Id": "bids::sub-01/func/sub-01_task-tonecounting_bold.nii",
+                "Label": "sub-01_task-tonecounting_bold.nii",
+                "AtLocation": "sub-01/func/sub-01_task-tonecounting_bold.nii",
```
i meant to use the bids uri here as well.

bclenet committed:
Using markdown mermaid graph for provenance

bclenet committed:
Scopes of provenance description

bclenet committed:
Environments added in provenance context

bclenet committed:
Removing non existing terms Entity Activity and Environment

bclenet committed:
Entities description

bclenet committed:
Descriptions for ids in metatadta

bclenet committed:
Descriptions for ids in metatadta

bclenet committed:
Removing caps to entity, activity, environment and software

bclenet committed:
descriptions of suffixes

cmaumet commented on src/schema/objects/suffixes.yaml:
```diff
@@ -502,6 +502,14 @@ VFA:
     ([Deoni et al. 2005](https://doi.org/10.1002/mrm.20314)).
+act:
+  value: act
+  display_name: Provenance activity records
```
Let's simplify in "Activity" to be consistent with the definitions at the top of the spec: https://bids-specification--2099.org.readthedocs.build/en/2099/modality-agnostic-files/provenance.html#definitions

cmaumet commented on src/schema/objects/suffixes.yaml:
```diff
@@ -612,6 +620,22 @@ electrodes:
     File that gives the location of (i)EEG electrodes.
+ent:
+  value: ent
+  display_name: Provenance entity records
```
--> "Entities"

cmaumet commented on src/schema/objects/suffixes.yaml:
```diff
@@ -612,6 +620,22 @@ electrodes:
+    See the [`Entities`](SPEC_ROOT/modality-agnostic-files/provenance.md#entities) section.
+env:
+  value: env
+  display_name: Provenance environment records
```
--> "Environnements"

cmaumet commented on src/schema/objects/suffixes.yaml:
```diff
@@ -817,6 +841,14 @@ probseg:
     This suffix may only be used in derivative datasets.
+soft:
+  value: soft
+  display_name: Provenance software records
```
--> "Software"

cmaumet commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,496 @@
+-->
+{{ MACROS___make_subobject_table("metadata.GeneratedBy.items") }}
+
+## Provenance files
```
Camille stopped her review here ----

cmaumet reviewed (commented):
Hi @bclenet !

Thanks for this version. I've included a number of comments (mainly some rephrasing).

bclenet committed:
@cmaumet review

bclenet committed:
@cmaumet review

bclenet commented on src/modality-agnostic-files/provenance.md:
````diff
@@ -0,0 +1,496 @@
+```
+
+!!! Note
+    Files sharing the same `prov` BIDS entity contain provenance records that belong to the same provenance group.
````
As discussed, we will remove the definition for provenance group .

bclenet committed:
GeneratedBy can be an array of objects (pipelines) or an activitiy Id

bclenet committed:
Removing groups of provenance

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bclenet commented on src/schema/objects/metadata.yaml:
```diff
@@ -1379,23 +1555,81 @@ GeneratedBy:
+          RECOMMENDED if `Name` is `"Manual"`.
+        type: string
         $ref: objects.metadata.Description__GeneratedBy
       CodeURL:
```
I reverted the description for `GeneratedBy` in `dataset_description.json` to:

> Used to specify provenance of the dataset.

And detailed the two options for this field in the spec (Provenance of a BIDS dataset section)

bclenet committed:
GeneratedBy schema modification

bclenet committed:
Reorganize and consistency in naming

bclenet committed:
Using identifier instead of

bclenet committed:
Id for entities other than BIDS URI

bclenet committed:
Minimal example for study dataset

cmaumet reviewed (commented):
Hi Boris, 

Thanks a lot for this new version of the BIDS-Prov spec.

As discussed by email I reviewed the "key concepts" section as well as the outline.

Overall, it looks great and I think the outline gives a good flow of information.

One suggestion would be :

----
Overview
---- Goals
---- General principles
---- Key concepts
~~Provenance files~~
---- ~~Activities~~
---- ~~Entities~~
---- ~~Software~~
---- ~~Environments~~
Provenance of a BIDS file
---- Sidecar json
---- Provenance files
-------- Activities
-------- Software
-------- Environments
-------- Entities
Provenance of a BIDS dataset
---- Description using provenance records
---- Description of processes or pipelines
Consistency and uniqueness of identifiers
---- Identifiers for entities
---- Identifiers for other provenance records
Minimal examples
---- Provenance of a BIDS raw dataset
---- Provenance of a BIDS study dataset

----

And the section "Provenance of a BIDS dataset" would refer back to the subection "Provenance files" as needed.

bclenet committed:
Removing provenance metadata

bclenet committed:
Provenance file organization

satra commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,496 @@
+) }}
+
+!!! note
+    Not defining the `SidecarGeneratedBy` field means that the sidecar JSON was generated by the same activity as the data file itself and described in the `GenearatedBy` field.
```
there may be some dual implicitness in the `GeneratedBy` key implying both file+sidecar. so either it should be clarified in the spec or as camille says it should be separated through this notion of the new keyword `SidecarGeneratedBy`

bclenet committed:
Removing note about SidecarGeneratedBy

bclenet committed:
prov BIDS entity description

effigies committed:
Update metaschema

effigies committed:
Merge remote-tracking branch 'upstream/master' into BEP028_spec

effigies commented on mkdocs.yml:
```diff
@@ -120,6 +121,11 @@ markdown_extensions:
         - name: tsvgz
           class: tsv
           format: !!python/name:bidsschematools.render.tsv.fence
+  - pymdownx.superfences:
+      custom_fences:
```
```suggestion
```

effigies committed:
Update mkdocs.yml

effigies committed:
fix metaschema

effigies committed:
chore: codespell -w

effigies commented:
Here are the outstanding issues I can see, given the current state of the pull requests:

1) `SIDECAR_WITHOUT_DATAFILE` errors. I have pointed @bclenet to where this would need to be addressed: https://github.com/bids-standard/bids-validator/blob/c331a16/src/validators/internal/unusedFile.ts. The basic problem is that JSON files are not generally distinguishable from sidecars, which we do want to check apply to some file. We check against the two exceptions to this rule, but that list inclusion check could be made a function to handle prov.
2) Unvalidated prov contents. Use `schema/rules/json/prov.yaml` to define the contents of the files. I hope the existing files will be a good model, but let me know if you need some scaffolding there.
3) The arbitrary subdirectories are going to be a pain. The simplest thing would be to drop the idea and have a flat directory, which works right now. The alternative is probably going to involve new concepts in both the file rules of the directory rules (cc @rwblair): https://github.com/bids-standard/bids-specification/blob/aa2dda5b2f279606de450e8d47d4a36f02981e03/src/schema/rules/files/common/modality_agnostic.yaml#L5-L17 https://github.com/bids-standard/bids-specification/blob/aa2dda5b2f279606de450e8d47d4a36f02981e03/src/schema/rules/directories.yaml#L48-L109

cmaumet reviewed (commented):
This is another set of proposed updates as per our discussions

bclenet commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,697 @@
+    - [Provenance records for fMRI preprocessing using `fMRIPrep`](https://github.com/bclenet/bids-examples/tree/BEP028_fmriprep/provenance_fmriprep) - [Associated Pull Request#502](https://github.com/bids-standard/bids-examples/pull/502)
+
+    Further datasets are available from
+    the [BIDS examples repository](https://bids-website.readthedocs.io/en/latest/datasets/examples.html#provenance).
```
Yes, although I'm not sure why the table is already on the website :thinking:

bclenet committed:
Removing provenance records

bclenet committed:
Yaml lint

bclenet commented on src/schema/objects/metadata.yaml:
```diff
@@ -1176,6 +1253,104 @@ EncodingTechnique:
+      AtLocation:
+        name: AtLocation
+        description: |
+          For input files, this is the relative path to the file on disk.
```
use data instead of file

effigies commented:
I had a chance to talk with @rwblair about the arbitrary subdirectories yesterday. At least from the examples given (`/prov/preprocspm/prov-preprocspm{1,2}_*.json`), the idea is to allow grouping, which BIDS generally does with entities. Why not:

```
/prov[/prov-<label>]/prov-<label>_desc-<label>_<suffix>.json
```

Although a bit different from how BIDS has done things (data type before first entity directory), the machinery we already have in the schema is sufficient to encode this and the changes to the validator should not be difficult.

bclenet committed:
GeneratedBy is always an array

bclenet committed:
GeneratedBy is always an array

yarikoptic commented:
> ```
> /prov[/prov-<label>]/prov-<label>_desc-<label>_<suffix>.json
> ```
> 
> Although a bit different from how BIDS has done things (data type before first entity directory),

FWIW, I like it since it is generic.  Might also be applicable to e.g. [BEP044:Stimuli](https://bids.neuroimaging.io/extensions/beps/bep_044.html) where ATM stimuli files are not groupped but have `stim-<label>_...otherentities` flat listing.  

But overall it comes to the question when is worth keeping flat vs creating those folders, and it is kinda a generic aspect: e.g. if there is a BIDS dataset with only T1w images for 10 subjects -- folders are not really making it easier to navigate the data. Somewhat of a usecase for 

- https://github.com/bids-standard/bids-specification/pull/1809 .

Similarly here: if it is just a single "stage" derivative dataset (e.g. bids-app applied in one go across all subjects) -- there is no need for subfolders there, right?

effigies commented:
I have no objections to keeping it flat. I was under the impression that nesting was important.

bclenet committed:
Replacing Entities with ProvEntities

bclenet committed:
Replacing entities with provEntities in schema and context

bclenet committed:
New template for provenance files

bclenet committed:
Provenance description file

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bclenet committed:
Dataset names

bclenet committed:
Templates for files in schema.rules.files.common

bclenet committed:
Key concepts

bclenet committed:
provEntities for files and datasets are allowed in _ent provenance files

bclenet committed:
Light formatting

bclenet committed:
Allow fragments in BIDS URIs

bclenet committed:
provenance.tsv is under prov/

bclenet committed:
Fragments and BIDS URIs for datasets

bclenet committed:
Add first version of RDF paragraph

bclenet committed:
Moving provenance-context to src/

bclenet committed:
Codespell

bclenet committed:
Remark

bclenet committed:
Schema rules for provenance files

cmaumet committed:
Update src/modality-agnostic-files/provenance.md

Include comment for ProvEntity versus Entity

bclenet committed:
GeneratedBy allowing single id

bclenet committed:
GeneratedBy allowing single id

bclenet committed:
Checksum functions

bclenet committed:
Checksum functions

bclenet commented:
Hi @rwblair,

As discussed earlier, here are the issues I'm currently facing:

- **schema** | validation of provenance files: I created the `schema/rules/json/prov.yaml` file; are there other things to do in this files for validating the contents of the objects in the Activities (resp. Environments, ProvEntities, Software) arrays ?
- **validator** | .json provenance files are considered as sidecars, hence raising the error : SIDECAR_WITHOUT_DATAFILE -> i'm afraid i'll not be able to come up with a clean modification of the validator to mark these files as standalone json.
- **schema** | how/where to allow for a `prov/[prov-<label>]` directory in the schema ?
- **macros** | how to add an optional `prov/[prov-<label>]` in the filename templates of the *Provenance Files* section (using the `MACROS___make_filename_template` macro)
- **CI** | pdf rendering fails due to a `assert_no_multiline_links` which I don't understand

Thanks for your help :)

bclenet committed:
generatedby disambiguation inside dataset_description

bclenet committed:
generatedby disambiguation inside dataset_description

rwblair commented on src/schema/objects/metadata.yaml:
```diff
@@ -823,6 +896,17 @@ DewarPosition:
+    `MD5`; `SHA1`; `SHA-224` ; `SHA-256` ; `SHA-384` ; `SHA-512` ;
+    `SHA3-224`; `SHA3-256`; `SHA3-384`; `SHA3-512`; `BLAKE2B-256`; `BLAKE3-256`;
+    `SHAKE128`; `SHAKE256`. Otherwise, key MAY be an arbitrary label.
+    The corresponding value is the checksum as computed by the function.
```
```suggestion
    The corresponding value is the checksum as computed by the function identified by the key.
```

rwblair commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,738 @@
+    Further datasets are available from
+    the [BIDS examples repository](https://bids-website.readthedocs.io/en/latest/datasets/examples.html#provenance).
+
+## Overview
```
I'd remove this heading and move the sentences under it to the above section (#provenance), I think the opening sentences of a page are generally understood to be an overview.

rwblair commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,738 @@
+
+This description is based on the [W3C Prov](https://www.w3.org/TR/2013/REC-prov-o-20130430/) standard.
+
+### General principles
```
These principles could be moved into the top level section, other sections of the standard (behavioral, phenotypic, etc) use requirement keywords in their opening salvos. I'd also remove the new lines between the sentences.

bclenet committed:
Context update

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bclenet committed:
Context update

bclenet committed:
GeneratedBy issues

bclenet committed:
Minimal examples

bclenet committed:
Minimal example descriptions

rwblair commented:
@bclenet While thinking about the first two points on how to validate certain parts of the files I started to realize the nature of the changes that will be required of the schema to validate them.

Here is my understanding of the main rules that this bep wants to enforce that require information outside of what has historically been used for validation, and the issues they raise.:

- provenance.tsv "provenance_label": There exists at least one file in the dataset that uses each value in the column for its prov- entity  label.
  - This is similar to the rule we have for participants.tsv that states any value in participant_id column must have a corresponding files in the dataset or be referenced in the phenotype directory. For this to work we had to populate a special field in the dataset context that had information about the names of the subject directories. Populating these fields is not representable in the Schema presently and must be implemented by the validators.
- Every id used in all prov files is unique with respect to the dataset.
  - When a given file is validated we build a context for it that has all the information it needs to be validated by the schema. Typically this involves loading a sidecar or a handful of specific associated files. This type of pan-dataset assertion is not possible in the current schema.
- *_ent.json "locatedAt": Its value exists in the dataset.
  - Doable in the current schema.

The following are all similar:
- *_ent.json "generatedBy": Its values only reference existing Activity ids
- *_act.json "AssociatedWith": Its Values only reference existing Software ids.
- *_act.json "Used": Its Values in references only reference existing Environment ids or ProvEntity Ids
- anybidsfile.json "GeneratedBy": Its Values only reference existing Activity ids
- anyBidsfile.json "SidecarGeneratedBy": Its Values only reference existing Activty ids
  - This suffers from the above issue. For any given prov file we are validating we must load arbitrarily many others and check the value at a specific key inside each of their arrays. We could try and come up with new semantics for schema entries that would allow this. Another option would be to come up with a new way in the schema to aggregate values from multiple files into a single place, and then a way of running checks on the aggregated data. But, even if all the, for example, Software objects were gathered in a single place we'd have another problem...
  - The expression language is incapable of iterating through an array of objects and running a check on a specific key for each element. We could extend the expression language to have a function like `flatten(array: List[List  | dict], key: Optional[star]`. If the input is an array of arrays then we use normal flatten semantics of putting all elements of all arrays in a single array and return that. If the input is an Array of objects we return an array of each objects value at the `key` specified.
 
One thing I like about this proposal is that each json file is simple enough to be immediately understood by a human. I was playing around with alternative ways of organizing data from the examples that might be more amenable to the current expression language and they were all much more difficult to read at a glance. The UID in the Ids makes me think this was not meant to be produced or consumed by humans, but I'm a sucker for looking at any json file that comes across my path.
 
Please let me know if I have misunderstood/misinterpreted any of the rules from the BEP.
 
@effigies Any comments on my characterizations of the schema's short comings with respect to the above rules?

@bclenet This only sort of addressed your first two issues, for the remaining four:
- sidecar without datafile - We do need to add a field to the schema to indicate this, next schema hack I'll bring it up, and take ownership of adding its interpretation to the javascript validator.
- prov subdirectories - I've got a local branch that's capable of doing this, I'll try and push it upstream when its ready.
- Macros and CI - I still need to look into these.

bclenet committed:
Update src/schema/objects/metadata.yaml

Co-authored-by: Ross Blair <rosswilsonblair@gmail.com>

bclenet committed:
Update src/schema/objects/metadata.yaml

Co-authored-by: Ross Blair <rosswilsonblair@gmail.com>

bclenet committed:
Update src/modality-agnostic-files/provenance.md

Co-authored-by: Ross Blair <rosswilsonblair@gmail.com>

bclenet committed:
Ids for provEntities

bclenet committed:
Consistency of metadata fields description

bclenet committed:
RDF section

bclenet committed:
RDF section

bclenet committed:
Study dataset minimal example

bclenet committed:
Codespell

bclenet committed:
Possibility to set Command metadata to null for manual activities

cmaumet committed:
shorten example text and link to the BIDS website

cmaumet committed:
Note as text instead of a box (this info is more of some additional info rather than an important note

cmaumet committed:
condense the intro

cmaumet committed:
link to subsection RDF

cmaumet committed:
removing heading as per https://github.com/bids-standard/bids-specification/pull/2099#discussion_r2437268020

cmaumet committed:
text formatting + remove mention of metadata fields

cmaumet committed:
use extension rather than 'one or more'

cmaumet committed:
fix typo

bclenet committed:
JSON null as type

cmaumet committed:
Update link to examples

cmaumet committed:
Update metadata fields info

bclenet committed:
Merge pull request #1 from cmaumet/BEP028_spec

[ENH] Update to BIDS-Prov specification pull request

bclenet committed:
provenance.tsv rules

bclenet committed:
Activity/activities

bclenet committed:
Provenance page reorganize

bclenet committed:
Merge branch 'master' into BEP028_spec

bclenet committed:
Moving provenance graph to RDF section

bclenet committed:
Provemance context adress

bclenet committed:
Rephrase intros for sections Provenance of...

bclenet committed:
Removing notion of provenance objects

bclenet committed:
Removing notion of provenance objects

bclenet committed:
Removing notion of provenance objects

bclenet committed:
Rules for provenance.tsv

bclenet committed:
Merge branch 'master' into BEP028_spec

cmaumet committed:
fix link

cmaumet commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,771 @@
+    - `bids:ds001734:sub-002/anat/sub-02_T1w.nii` - a provEntity describing a T1w file for subject `sub-002` in the `ds001734` dataset;
+    - `bids::sub-014/func/sub-014_task-MGT_run-01_events.tsv` - a provEntity describing an events file for subject `sub-014` in the current dataset;
+    - `bids:fmriprep:sub-001/func/sub-001_task-MGT_run-01_bold_space-MNI152NLin2009cAsym_preproc.nii.gz` - a provEntity describing a bold file for subject `sub-001` in the `fmriprep` dataset;
+    - `bids::prov#provEntity-acea8093` - a provEntity describing a file that is not available in the dataset.
```
```suggestion
    - `bids::prov#provEntity-acea8093` - identifier for a file that is not available in the dataset.
```

cmaumet commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,771 @@
+## Provenance from an RDF perspective
+
+!!! note
+    The [Resource Description Framework (RDF)](https://www.w3.org/RDF/) is a method to describe and exchange graph data.
```
I think we could remove this note and instead add a parenthesis after "RDF graph" below. See proposed edit

cmaumet reviewed (commented):
@bclenet - thanks a lot for this new version! See some small edits and suggestions.

cmaumet committed:
Update src/modality-agnostic-files/provenance.md

bclenet committed:
Update src/schema/objects/metadata.yaml

Co-authored-by: Camille Maumet <cmaumet@users.noreply.github.com>

bclenet commented on src/schema/objects/metadata.yaml:
```diff
@@ -823,6 +900,17 @@ DewarPosition:
+  name: Digest
+  display_name: Digest
+  description: |
+    Object containing digests of the data described by the provEntity.
```
@cmaumet Actually, this is something I wanted to discuss with you ; because it could be nice to allow `Digest` and `Type` in `dataset_description.json` as well. What do you think about that ?

bclenet committed:
Consistency of identifiers

bclenet commented on src/schema/objects/metadata.yaml:
```diff
@@ -1396,6 +1581,22 @@ GeneratedBy:
         $ref: objects.metadata.CodeURL
       Container:
         $ref: objects.metadata.Container
```
This comes from the master branch.
https://bids-specification.readthedocs.io/en/stable/modality-agnostic-files/dataset-description.html#dataset_descriptionjson

I'll think about how / if we should change it.

bclenet committed:
cmaumet review

bclenet committed:
ProvEntities are files

bclenet committed:
Identifier for

bclenet committed:
Description for Type

bclenet committed:
Trying to remove occurences of provEntity

bclenet committed:
More precise descriptions for suffixes

bclenet committed:
yamllint

bclenet committed:
Back to provenance Type having its own metadata

bclenet committed:
Files and Datasets as terms

bclenet committed:
Removing ProvEntities

bclenet committed:
Optional subdirectories under prov/

bclenet committed:
Rule for all sidecars to contain provenance fields

bclenet committed:
dataset_description generatedby

bclenet committed:
yamlint

bclenet committed:
Selectors

bclenet committed:
Selectors

bclenet committed:
Selectors

bclenet committed:
Selectors

bclenet committed:
Add empty line after amonition block declaration

bclenet committed:
Merge branch 'master' into BEP028_spec

bclenet committed:
Add empty line after amonition block declaration

bclenet committed:
Double_ in metadata name for dataset generatedby

bclenet committed:
refactor metadata fields

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bclenet committed:
refactor metadata fields

bclenet committed:
refactor metadata fields

bclenet committed:
refactor metadata fields

bclenet committed:
Merge branch 'master' into BEP028_spec

cmaumet committed:
omit 'section' for internal links

cmaumet committed:
rephrase intro

cmaumet committed:
link to ex

cmaumet committed:
text shortening

cmaumet committed:
shorten description generatedby for datasets

cmaumet committed:
text link to ex

cmaumet committed:
remove tool mention act

cmaumet committed:
change all snipped to for a complete ex

cmaumet committed:
RRID or other alternative id

cmaumet committed:
acted on behalf desc

bclenet committed:
provEntity sentance

bclenet committed:
Merge branch 'master' into BEP028_spec

bclenet committed:
Merge pull request #2 from cmaumet/BEP028_spec

Proofreading small updates

bclenet committed:
Other comments from Camille

bclenet committed:
Removing desc entity for provenance files

bclenet committed:
Consistncy with SPM example

bclenet committed:
Entities -> prov:Entity

bclenet committed:
Entities -> prov:Entity

bclenet committed:
prov:Entities -> prov:Entity

bclenet committed:
URI consitency with examples

yarikoptic commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,818 @@
+        "Software": [
+            {
+                "Id": "bids::prov#dcm2niix-khhkm7u1",
+                "AltIdentifier": "RRID:SCR_023517",
```
I would right away

```suggestion
                "AltIdentifiers": ["RRID:SCR_023517"],
```

not sure which one to add here, has no FairSharing one etc

yarikoptic reviewed (changes requested):
initial ones... need to look more

bclenet committed:
Review from Yarik and Camille

Co-authored-by: Yaroslav Halchenko <debian@onerussian.com>
Co-authored-by: Camille Maumet <cmaumet@users.noreply.github.com>

bclenet committed:
NIfTI

bclenet committed:
Description for activities

bclenet committed:
Description for activities

bclenet committed:
Schema update

bclenet committed:
Update spec with manual example

bclenet committed:
altidentifier as a list

bclenet committed:
provenance_label > provenance_id

bclenet committed:
provenance_id

bclenet committed:
All PROV-O Starting Point Terms in context

bclenet committed:
Merge branch 'master' into BEP028_spec

bclenet committed:
External file description not needed

satra reviewed (approved):
looks great. just a small comment on removing a sentence.

cmaumet commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,822 @@
+
+    Several [example BIDS-Prov datasets](https://bids-website.readthedocs.io/en/latest/datasets/examples.html#provenance) have been formatted using this specification and can be used for practical guidance when curating a new dataset.
+
+This part of the BIDS specification is aimed at describing the provenance of a BIDS dataset. This description is retrospective: it describes a set of steps that were executed in order to establish the dataset and is based on [W3C Prov](https://www.w3.org/TR/2013/REC-prov-o-20130430/) (see [Provenance from an RDF perspective](#provenance-from-a-rdf-perspective)). Note: this is different from prosp ...
```
Thanks for your review and for catching this @satra - I second removing the note.

bclenet committed:
Satra review

bclenet committed:
Merge branch 'bids-standard:master' into BEP028_spec

bclenet committed:
Example names

effigies commented on src/schema/objects/metadata.yaml:
```diff
@@ -4378,6 +4766,34 @@ Type:
+  display_name: Type
+  description: |
+    Term(s) from (a) controlled vocabulary/vocabularies that more specifically
+    describes the file.
```
This is unclear. I think it means that this is `prov:type`, but I don't think many readers will understand this.

effigies commented on src/schema/objects/metadata.yaml:
```diff
@@ -3889,6 +4228,55 @@ SliceTiming:
+          Example: if software A launches software B to perform activity C, then B ActedOnBehalfOf A.
+
+          Related software MUST be described as specified in the
+          [Software](SPEC_ROOT/modality-agnostic-files/provenance.md#software) section.
```
This MUST is unenforced, correct?

effigies commented on src/schema/rules/checks/provenance.yaml:
```diff
@@ -0,0 +1,19 @@
+    level: error
+  selectors:
+    - datatype == 'provenance'
+    - suffix == '.tsv'
```
```suggestion
    - suffix == 'provenance'
    - extension == '.tsv'
```

effigies reviewed (commented):
I have a number of narrow suggestions/questions/comments below. Here are some broader thoughts:

1) `Id` and `Label` keep showing up, but it feels like it would be useful to introduce them generally as a unique identifier and a human-friendly name.
2) When we have direct correspondences with RDF/Prov/XSD/etc terms, we should note that in the term descriptions, similar to what we do with DICOM. That way they show up in the glossary.
3) We should strive not to use `make_metadata_table` more than necessary. If it's worth rendering procedurally, it's probably worth having in the schema.
4) I think using containers is going to be the most common kind of environment people are interested in, so it would be good to show an example. (The fMRIPrep example has it. Which reminds me, shouldn't environments be able to have digests?)
5) There are a lot of things that are strings or arrays of strings. Given that most provenance is going to be written automatically, this flexibility doesn't seem worth it to me. I would suggest making them all arrays.
1) It would be helpful to adhere to https://sembr.org/. (See also https://github.com/bids-standard/bids-specification/blob/master/CONTRIBUTING.md#soft-rules)

rwblair commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,822 @@
+
+## Provenance of a BIDS file
+
+Provenance of a BIDS file SHOULD be stored inside its sidecar JSON.
```
```suggestion
Provenance of a BIDS data file SHOULD be stored inside its sidecar JSON.
```

Specifying data files since things like dataset_description.json don't have sidecars but also aren't a sidecar to any other file.

rwblair commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,822 @@
+
+### Description using identifiers
+
+This section details how to describe provenance of a dataset using identifiers.
```
```suggestion
This section details how to describe provenance of a dataset using identifiers.
The following fields are intended for use in `dataset_description.json` to provide provenance information that applies to the entire dataset.
```

rwblair commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,822 @@
+
+    These files SHOULD not describe the current dataset. See [Provenance of a BIDS dataset](#provenance-of-a-bids-dataset) for this purpose.
+
+Each file MUST include at least one of the following key:
```
```suggestion
Each file MUST include one or more of the following keys:
```

rwblair reviewed (commented):
I second the points from @effigies comment. The first two seem about, introductory material for `ID` and `Label` along with linking to external ontology terms/definitions when applicable, seem particularly useful.

bclenet committed:
Review from Ross and Chris

Co-authored-by: Chris Markiewicz <effigies@gmail.com>
Co-authored-by: Ross Blair <rosswilsonblair@gmail.com>

bclenet committed:
Review from Ross and Chris, part 2

Co-authored-by: Chris Markiewicz <effigies@gmail.com>
Co-authored-by: Ross Blair <rosswilsonblair@gmail.com>

bclenet committed:
Review from Ross and Chris, part 3

Co-authored-by: Ross Blair <rosswilsonblair@gmail.com>
Co-authored-by: Chris Markiewicz <effigies@gmail.com>

bclenet commented on src/common-principles.md:
```diff
@@ -967,18 +967,24 @@ The `dataset-name` component is an identifier for a BIDS dataset,
 bids::sub-01/fmap/sub-01_dir-AP_epi.nii.gz
 bids:ds000001:sub-02/anat/sub-02_T1w.nii.gz
 bids:myderivatives:sub-03/func/sub-03_task-rest_space-MNI152_bold.nii.gz
+bids:fmriprep:sub-001/anat/sub-001_T1w_preproc.nii.gz#aa56ztg8
```
I'm not against removing this example, but do you think it would be clearer if we explain what this id is intended for ?

markmikkelsen replied:
I agree. I was reviewing the BEP and wondered if there is a cosmetically better way to generate example `<uid>`.

bclenet committed:
Review from Ross and Chris, part 4

bclenet committed:
Sembr / soft rules

bclenet committed:
entity prov -> provenance

satra commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,822 @@
+        "Software": [
+            {
+                "Id": "bids::prov#dcm2niix-khhkm7u1",
+                "AltIdentifier": ["RRID:SCR_023517"],
```
for software, the reason a custom id is required here, or some mutilation of RRIDs is that RRIDs are not version specific, and since in a graph sense the `Id` is the node to which properties are attached, if an RRID is used then the edges from that node would conflate all software versions used with the `Id`. if any RRID or doi is version specific, then of course that can be used as `Id`. 

`Id`'s should not merge entities unless they are the "same" from an information standpoint. for software, i would argue that provenance about version is critical and should be treated as different software agents, hence requiring different `Id`s. ideally two people using the same software version should generate the same `Id` but that would require some coordination, whereas matching could be done post-hoc.

bclenet committed:
Merge branch 'master' into BEP028_spec

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bclenet committed:
remark

bclenet committed:
schema provenance entity

effigies commented on src/schema/rules/json/prov.yaml:
```diff
@@ -0,0 +1,63 @@
+    Files:
+      level: optional
+      level_addendum: required if `prov:Entity` and `Datasets` fields are absent
+    Dataset:
```
```suggestion
    Datasets:
```

effigies reviewed (commented):
Two quick fixes.

effigies committed:
Apply suggestions from code review

Co-authored-by: Ross Blair <rosswilsonblair@gmail.com>

effigies committed:
Update src/schema/rules/json/prov.yaml

dependabot[bot] committed:
chore(deps): bump prettier in the node-utilities group (#2334)

Bumps the node-utilities group with 1 update: [prettier](https://github.com/prettier/prettier).


Updates `prettier` from 3.7.4 to 3.8.0
- [Release notes](https://github.com/prettier/prettier/releases)
- [Changelog](https://github.com/prettier/prettier/blob/main/CHANGELOG.md)
- [Commits](https://github.com/prettier/prettier/compare/3.7.4...3.8.0)

---
updated-dependencies:
- dependency-name: prettier
  dependency-version: 3.8.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
  dependency-group: node-utilities
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

bclenet committed:
BIDS compliant RDF graph

bclenet committed:
Check provenance.tsv is inside prov/

bclenet committed:
Correspondances with WC3 prov

bclenet committed:
AlternativeIdentifier

bclenet committed:
type: array yaml errors

bclenet committed:
yamlint

bclenet committed:
Correspondances for Type

bclenet committed:
Digest for Environments

bclenet committed:
Id correspondance

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bclenet committed:
JSON in markdown

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bclenet committed:
Id correspondance

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bclenet committed:
yamlint

bclenet committed:
AtLocation correspondance

bclenet committed:
rdf:type instead of json-ld:@type

bclenet committed:
Rm digets and type for dataset

bclenet committed:
Rm Digest and Type for Dataset

bclenet committed:
Use of Type in RDF section

bclenet committed:
Use of Type in RDF section

bclenet committed:
Environment example

bclenet commented on src/schema/objects/metadata.yaml:
```diff
@@ -929,6 +1054,28 @@ DewarPosition:
+    `SHA3-224`; `SHA3-256`; `SHA3-384`; `SHA3-512`; `BLAKE2B-256`; `BLAKE3-256`;
+    `SHAKE128`; `SHAKE256`. Otherwise, key MAY be an arbitrary label.
+    The corresponding value is the checksum as computed by the function identified by the key.
+  type: object
```
We decided to remove Digest for datasets, as we currently have no procedure for generating whole dataset digests.

bclenet committed:
Citing BIDS-Prov

bclenet committed:
Camille review

bclenet committed:
Camille review

bclenet committed:
Camille review

cmaumet commented on src/modality-agnostic-files/provenance.md:
```diff
@@ -0,0 +1,904 @@
+Provenance information reflects the provenance of a full dataset
+and/or of specific files at any level of the BIDS hierarchy.
+Provenance information SHOULD not include human subject identifying data.
+
```
Hi @rwblair @effigies @yarikoptic, thx for the suggestion! Just discussed with @bclenet and we think we could go with:
 - Renaming the `Label` in BIDS-Prov into something else, e.g. `Name` so as to avoid confusion with bids `label` 
  - and then the update would be:
 
```suggestion

!!! Note
Throughout this document, the terms `Id` and `Name` are used to provide identification for JSON objects related to provenance.
`Id` is used to unambiguously [identify those objects](#Provenance-identifiers) that may be referenced elsewhere,
permitting automated tools to [construct and query a graph](#Provenance-graph).
`Name` is a human-readable name for that object, which need not be unique.
```

How is that?

bclenet committed:
Provenance identifiers & provenance graph

bclenet committed:
Identifier(s) links to Provenance identifiers section

yarikoptic commented on src/schema/objects/metadata.yaml:
```diff
@@ -67,6 +67,101 @@ AcquisitionVoxelSize:
+        description: |
+          Command (or commands) performed by the activity, including all parameters.
+
+          Set to `null` to describe that the activity was performed manually.
```
I love present alignments to `rdfs:` ... is there any property within LD PROV ecosystem which could relate here?

yarikoptic commented on src/schema/objects/metadata.yaml:
```diff
@@ -67,6 +67,101 @@ AcquisitionVoxelSize:
+        description: |
+          Name for the activity.
+
+          Corresponds to [RDF Schema `rdfs:label`](https://www.w3.org/TR/rdf-schema/#ch_label).
```
this and other definitions (Id etc) are used multiple times across different metadata "array" properties. Looking through here is an example of how to define each once on top level and then reuse in specific constructs -- @effigies tell me if I am misguiding here:

```yaml
DeidentificationMethodCodeSequence:
  name: DeidentificationMethodCodeSequence
  display_name: Deidentification Method Code Sequence
  description: |
    A sequence of code objects describing the mechanism or method use to remove the Patient's identity.
    Corresponds to [DICOM Tag 0012, 0064](https://dicomlookup.com/dicomtags/(0012,0064))
    `De-identification Method Code Sequence`.
  type: array
  items:
    type: object
    recommended:
      - CodeValue
      - CodeMeaning
      - CodingSchemeDesignator
      - CodingSchemeVersion
    properties:
      CodeValue:
        $ref: objects.metadata.CodeValue
      CodeMeaning:
        $ref: objects.metadata.CodeMeaning
      CodingSchemeDesignator:
        $ref: objects.metadata.CodingSchemeDesignator
      CodingSchemeVersion:
...
```

it should be quite doable by claude-code to achieve quickly, let me know if I should push such RF

bclenet committed:
Activities.Description is an extended description

Co-authored-by: Yaroslav Halchenko <debian@onerussian.com>

bclenet committed:
Activities.Description -> rdfs:comment

bclenet committed:
Clarification note about Id and Label

bclenet committed:
yamlint

rwblair committed:
have subprov use full entity keyname 'provenance' to allow proper lookup in schema.objects.entities

markmikkelsen commented on src/schema/objects/suffixes.yaml:
```diff
@@ -620,7 +626,20 @@ electrodes:
+  description: |
+    A JSON file containing objects describing  input and output data in the context of provenance.
+    (See the [`Input and output data`](SPEC_ROOT/modality-agnostic-files/provenance.md#-input-and-output-data)
+    section).
```
Instead of `ent`, wouldn't something like `io` make more sense?

julia-pfarr replied:
@markmikkelsen could you post this again in the https://github.com/bids-standard/bids-specification/discussions/2405? We're exclusively collecting the review of this BEP there now. Thank you!!

bclenet committed:
Digest -> Checksum using controlled terms

bclenet committed:
Alternative identifier for ChecksumAlgorithm

bclenet committed:
Alternative identifier for ChecksumAlgorithm

bclenet committed:
Typos in provenance.md

bclenet committed:
io suffix instead of ent

bclenet committed:
Typo in provenance context

bclenet committed:
ChecksumAlgorithm and Value in metadata

bclenet committed:
Typo in metadata

bclenet committed:
More examples for ChecksumAlgorithm

bclenet committed:
Store provenance of any file in sidecar when possible

bclenet committed:
may in uppercase
